### DISCUSSION TITLE
Using libcpr instead of libcurl

### DISCUSSION THREAD

**Discussion by R-Abbasi** [skunkforce/OmniView#148] [Ideas]
What do you think of using the libcpr rather than libcurl, please? 
Here's a quick _GET_ request:

```
#include <cpr/cpr.h>

int main(int argc, char** argv) {
    cpr::Response r = cpr::Get(cpr::Url{"https://api.github.com/repos/whoshuu/cpr/contributors"},
                      cpr::Authentication{"user", "pass", cpr::AuthMode::BASIC},
                      cpr::Parameters{{"anon", "true"}, {"key", "value"}});
    r.status_code;                  // 200
    r.header["content-type"];       // application/json; charset=utf-8
    r.text;                         // JSON text string
    return 0;
}
```
And here's [less functional, more complicated code, without cpr](https://gist.github.com/whoshuu/2dc858b8730079602044).

**Comment by MaxClerkwell** [comment]
`libcpr` provides a `C++`-interface for `libcurl`. So the implementation would be still `libcurl`. Anyhow, I guess, using a `C++`-frontend-wrapper looks like a good idea. 

Please provide an example of how to upload data to our api, using the following `libcurl`-wrappers:
* cpprestsdk
* libcpr
* Curlcpp
* libcurlpp

Also try them in a separate branch and figure out, which of these are most easily implementable into the CI.

**Comment by R-Abbasi** [comment]
Not much experienced with networking or those libraries yet but made something quickly. Sure. Will also try others in the middle of doing my tasks. 

This `libcpr` version is more or less (may need  modifications) doing the same work as [our libcurl version](https://github.com/skunkforce/OmniView/blob/feature-analyze-waveforms-2/src/sendData.hpp) that I tired well to simplify. Obviously, this version still looks much simpler. 
```
auto j = R"(
   {
     "meta": ["string"],
     "data": {
       "sampling_rate": 44100,
       "y_values": [0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20, ...]
       }
     }
  )"_json;
auto str = j.dump();
auto r = cpr::Post(cpr::Url{"https://api.auto-intern.de/check-request"},
              cpr::Body{str}, cpr::Header{{"Content-Type", "application/json"}});
std::cout << r.text << '\n'
```

> Also try them in a separate branch

Do you mean to test each in the project and then create a `pr` for them one-by-one, please?

**Comment by R-Abbasi** [comment]
Tested the code above using `CMakeLists.txt` below:
```
cmake_minimum_required(VERSION 3.20)
project(sendData)

set(CMAKE_CXX_STANDARD 23)
set(DEFAULT_BUILD_TYPE "Debug")
add_executable(sendData main.cpp)

if(WIN32) 
    set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
    set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
endif()

include(FetchContent)
FetchContent_Declare(cpr GIT_REPOSITORY https://github.com/libcpr/cpr.git
                         GIT_TAG 3020c34ae2b732121f37433e61599c34535e68a8) 
FetchContent_MakeAvailable(cpr)
target_link_libraries(sendData PRIVATE cpr::cpr)
```
It worked as expected on both Linux and Windows:

![2](https://github.com/skunkforce/OmniView/assets/53223982/2db3a019-65ca-47ec-b0b6-56d7e26aefe4)

Therefore, the current function `std::string sendData(const std::string &data)` can presumably be modified this way: 
```
#include <cpr/cpr.h>
#include <string_view>

inline std::string sendData(std::string_view data)
  auto r = cpr::Post(cpr::Url{"https://api.auto-intern.de/check-request"},
                     cpr::Body{data}, cpr::Header{{"Content-Type", "application/json"}});
  return  r.text; // response from the server
}
```
It's super simple with just few lines of code. We can supposedly even go further by removing the header-only library `sendData.hpp` wholly and putting that implementation right inside [generateTrainingData.cpp](https://github.com/skunkforce/OmniView/blob/feature-analyze-waveforms-2/src/generateTrainingData.cpp#L413). 
> which of these are most easily implementable

As shown, it merely contains the principles of an upload (the URL, Body and Header). I don't think we can make it simpler using any of the other libraries. But if you prefer, I'll try to test them and post their results too. @bjoekeldude
